Added paper details to Paper View

Slartibartfass2 opened the pull request:
Closes #41 

## What I have made

- The main change of this PR is that `PaperDetailsCard.svelte` now displays all the information, planned in the design
  - initially only the paper title, authors, year and publisher are shown
  - the publication type, publication name and external ID can be shown, when pressing 'show more information'
  - by default all inputs are made read-only, but there are two edit mode buttons to toggle that property
- For that functionality, I added a `ToggleableInput` component, which is an auto-sizing textarea
  - it's a textarea because it can handle multiple lines
  - it's auto-sizing because that's what common textareas can't do (we use an action for that)
- I made most paper related components async because the data won't be there instantly
  - all of them use skeletons while loading
- For a long time, I had the problem that the abstract textarea won't stop expanding when it reaches the end of the card, after trying many solutions with no success, I implemented a semi-hard-coded solution that works
- This solution is a svelte action applied to the textarea
  - on creation, on update, when input is added to the textarea or when the screen size changes the action is executed
  - the action calculates how much space is left between the top of the textarea and the end of the card as it's supposed to be, this distance is added as max-height attribute
  - when exceeding this height, a scrollbar is added to the abstract

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~[ ] unit tests~ no logic that be tested by a unit test added
  - [x] integration tests
  - ~[ ] end-to-end tests~ not available yet
- [ ] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add paper details to paper view

Slartibartfass2 committed:
feat: make paper view components async

Slartibartfass2 committed:
feat: add auto max height logic

Due to not finding any other solution I decided to write an action that automatically calculates the max height according to the screen height and the position of the text area.
This is updated whenever the screen size changes.

Slartibartfass2 committed:
feat: add catch cases to loading paper promises

Slartibartfass2 committed:
test: add component tests for paper details card

luca-schlecker commented on src/lib/components/composites/PaperBookmarkButton.svelte:
```diff
@@ -7,15 +7,17 @@
 
     interface Props {
-        paperId: number;
+        loadingPaperId: Promise<number>;
```
What's your reasoning for pulling that promise inside this component?
Couldn't this component be used as is when wrapped in svelte's `{#await ...}`?

This question arises multiple times throughout the code, one comment thread is enough though imo.

Slartibartfass2 replied:
I used this because the component does not heavily rely on the asynchronous loading, i.e., we can just display the component and as soon as the paper is loaded the functionality works as well.
Therefore, wrapping the component in an await block seemed too much to me.

Where else did you notice this?

luca-schlecker replied:
I guess the behavior you are describing is right. 👍 No need to change anything

luca-schlecker commented on src/lib/components/composites/input/ToggleableInput.svelte:
````diff
@@ -0,0 +1,49 @@
+```
+-->
+<textarea
+    use:autosize
+    use:inputAction={inputActionProps}
````
Do you think this workaround is necessary? The abstract's `textarea` could instead fill the remaining space. It would reduce complexity with the caveat of always filling the remaining space even when provided with a short text (see image below). Visually there wouldn't be too much of a difference with that one exception.

![image](https://github.com/user-attachments/assets/8683ca3a-237b-4848-b7bf-3f9b5b2cb5ff)

Slartibartfass2 replied:
What did you use to produce that?

I tried plenty of approaches and none worked when the text overflows. The action was some sort of last resort.

luca-schlecker replied:
The combination of
```css
flex-grow: 1;
flex-basis: 0;
overflow: scroll;
```
on the abstract.

Slartibartfass2 replied:
I couldn't reproduce this.
Can you apply the changes to a new branch where I can try it for myself?

luca-schlecker commented on src/lib/components/composites/paper-components/PaperInfo.svelte:
```diff
@@ -1,27 +1,36 @@
+{:then paper}
+    <div class="grid grid-flow-row gap-0">
+        <div class="flex flex-row gap-1 items-center truncate">
+            {#if "id" in paper}
+                <div class="w-fit text-default-sb-nc text-neutral-500">#{paper.id}</div>
+            {/if}
```
Is there a reason the `id` could be missing?

Slartibartfass2 replied:
Yes, this is used when the data is only a `PaperSpec` with no ID. This is the case on the paper create page.

luca-schlecker replied:
I see, makes sense. 👍

luca-schlecker commented on src/lib/components/composites/paper-components/paper-view/PaperDetail.svelte:
```diff
@@ -0,0 +1,40 @@
+
+    type Props = WithElementRef<HTMLAttributes<HTMLDivElement>> & {
+        key: string;
+        value: unknown;
```
As far as I can tell `value` is used for class names and content at the same time. This should probably be explained somewhere.

Slartibartfass2 replied:
Great catch! I added it to the component documentation.

luca-schlecker commented on src/lib/components/composites/paper-components/paper-view/cards/PaperDetailsCard.svelte:
```diff
@@ -1,11 +1,86 @@
+        })
+        .catch(() => {
+            basicInfos = { Title: "", Authors: "", Year: "", Publisher: "" };
+        });
```
I recognize a common pattern: Instead of relying on svelte's `{#await ...}` a state rune is used with a default value. In case you want to avoid `{#await ...}` (previous comment), I'd recommend simplifying that behavior similar to [what is shown here:](https://dev.to/jdgamble555/async-fetching-in-svelte-5-826)
```typescript
// resource.svelte.ts

export let resource = <T>(
    fn: () => Promise<T>,
    initialValue?: T
) => {

    const _rune = $state<{ value: T | undefined }>({
        value: initialValue
    });

    $effect(() => {
        fn().then((data) => {
            _rune.value = data;
        });
    });

    return _rune;
};
```

This could also be extended with a default error value. Maybe something like this:
```typescript
export let resource = <T>(
    fn: () => Promise<T>,
    initialValue?: T,
    onSuccess: (...) => T,
    onError?: (...) => T
) => { ... }
```

Slartibartfass2 replied:
Nice suggestion, I like it. It's way more clean and descriptive now.

luca-schlecker commented on src/lib/components/composites/paper-components/paper-view/cards/PaperDetailsCard.svelte:
```diff
@@ -14,24 +89,86 @@ Paper Card for paper details in the Paper View component.
+                <Skeleton class="flex h-[1.625rem] rounded-full w-[82%]" />
+                <Skeleton class="flex h-[1.625rem] rounded-full w-[50%]" />
+                <Skeleton class="flex h-[1.625rem] rounded-full w-[75%]" />
+                <Skeleton class="flex h-[1.625rem] rounded-full w-[90%]" />
```
```suggestion
                {#each [100,95,70,82,50,75,90] as width}
                    <Skeleton class="flex h-[1.625rem] rounded-full w-[{width}%]" />
                {/each}
```

Slartibartfass2 replied:
Good catch. I added it.

luca-schlecker commented on src/lib/components/composites/paper-components/paper-view/decision-buttons/DecisionButton.svelte:
```diff
@@ -8,16 +8,29 @@
         ...restProps
     }: Props = $props();
+
+    let paperId = $state<number | undefined>(undefined);
+    loadingPaperId.then((id) => (paperId = id)).catch(() => (paperId = undefined));
```
Why not a nullable id?

Slartibartfass2 replied:
Please elaborate.

luca-schlecker replied:
Nevermind, I thought the promise could just take a nullable id but that is just nonsense semantically.

luca-schlecker commented on tests/integration/paper-components/paper-view/paper-detail.test.ts:
```diff
@@ -0,0 +1,99 @@
+        expect(screen.queryByTestId("skeleton")).not.toBeNull();
+    });
+
+    test("When paper loading failed without error, then errot text is shown", async () => {
```
```suggestion
    test("When paper loading failed without error, then error text is shown", async () => {
```

Slartibartfass2 replied:
Good catch. Changed it.

luca-schlecker reviewed (changes requested):
The page looks great, nice work. :+1: 

I stumbled upon issues with the integration tests though, I attached some details in the form of a log.
The pipeline doesn't seem to encounter the same problem.

Due to the upcoming gRPC work, I did not review changes regarding controllers or backend APIs.

[log.txt](https://github.com/user-attachments/files/18496177/log.txt)

Slartibartfass2 committed:
chore: add missing documentation and fix typo

Slartibartfass2 committed:
refactor: use resource wrapper to handle asynchronous resource loading

Slartibartfass2 committed:
feat: remove promise rejection errors from being displayed

Those are mainly for the developer and don't provide any advantage for the user

Slartibartfass2 reviewed (commented):
Thank you for the review.


I couldn't reproduce the errors you got in that log. On which commit did you experience that?

luca-schlecker commented:
> I couldn't reproduce the errors you got in that log. On which commit did you experience that?

Will have to test it on the machine that produced the errors again. This machine works through the tests nicely. Either fixed or flakey.

Slartibartfass2 committed:
feat: add `IS_DEV_MODE` env variable

This enables to control some feature, we only want to display in the devlopment mode.
Currently, this only enables showing which media query is active.

Slartibartfass2 committed:
chore: use "dot" repoter and add test:no-e2e command

Slartibartfass2 committed:
feat: remove auto-size and max-height actions

They are replaced by a working flex configuration

Slartibartfass2 committed:
fix: remaining fixes after rebase

luca-schlecker reviewed (commented):
Looking great 👍. A rebase seems to be needed though.

luca-schlecker commented on src/lib/components/composites/PaperBookmarkButton.svelte:
```diff
@@ -7,15 +7,17 @@
     let isHovered = $state(false);
     const tooltipText = $derived(isBookmarked ? "Remove from Reading List" : "Add to Reading List");
+    let paperId: string | undefined = $state(undefined);
+    loadingPaperId.then((id) => (paperId = id)).catch(() => (paperId = undefined));
```
This could be a call to `resource`.

Slartibartfass2 replied:
Great catch, I changed it.

luca-schlecker commented on src/lib/components/composites/paper-components/paper-view/decision-buttons/DecisionButton.svelte:
```diff
@@ -8,16 +8,29 @@
         ...restProps
     }: Props = $props();
+
+    let paperId = $state<string | undefined>(undefined);
+    loadingPaperId.then((id) => (paperId = id)).catch(() => (paperId = undefined));
```
another call to `resource` here

Slartibartfass2 replied:
Great catch, I changed it.

luca-schlecker reviewed (changes requested):
Only these two minor things are left.

Slartibartfass2 committed:
refactor: use resource helper where possible

luca-schlecker reviewed (approved):
Looking great! 🚀
